### ISSUE TITLE
Job Runs should create derived Dataset objects

### ISSUE THREAD

**Issue by annehaley** [atlascope/atlascope#68]
When a Job Run is complete, a Dataset object of the type "derived" should also be created so that JobRun outputs can be used for further computation/visualization.

**Comment by annehaley** [comment]
The information obtained from a JobRun actually doesn't fit very well into the Dataset model (many fields would be non-applicable) and no information is gained by saving the outputs of a JobRun to a Dataset's `content`. In fact, information is obscured: the context of the JobRun (its inputs, script, etc) could not be saved directly to the Dataset object and would have to be nested inside a reverse relation to the original JobRun object. 

To simplify this, I propose we make the `derived_datasets` field on the Dataset model a ManyToMany of _JobRuns_, not other Datasets. We could change the name to `derived_analytics` or something similar. We can still use JobRuns as valid inputs for further computation, and we'll get to keep the context nearby.

**Comment by waxlamp** [comment]
Would it be possible for the job logic itself to create the `Dataset` object in the system? Then we can leave it to each job type to supply the logic to cause derived datasets to appear. Something feels off about have a `Dataset` link to `JobRun`s via a property called `derived_datasets`.

I'm happy to discuss this in person if needed.

**Comment by annehaley** [comment]
Originally, I had imagined that the JobRun object would create a Dataset object within the spawn function, but I realized we would have a lot of blank fields and missing context. I do think we would have to change the name of the field if we change the model to which it links. It may be best to discuss this, but I'm not available until Monday. How about during the whole team meeting?

**Comment by naglepuff** [comment]
Whatever the results of that discussion end up being, let's be sure to write it up as part of this thread (or elsewhere, as long as its accessible)? I think its important that we really nail down the relationship between `Dataset`, `derived_datasets`, and `JobRun` objects (and `Pin`s!). I think having well-defined concepts for these entities will really help with dicussing potential workflows in the system

**Comment by annehaley** [comment]
In a discussion this evening, we decided that JobRuns can still create Dataset objects from their outputs. The JobRun model will need a new field to reference the newly-created Dataset, and the Dataset serializer will need a field to reference the reverse relation to the JobRun that created it.

**Comment by annehaley** [comment]
During our pair programming today, @marySalvi and I realized that the information held by the `JobRun` object can be contained in the `content` of a Dataset, which means that when we spawn a job and create these derived Dataset objects, we don't actually need the "middle man" JobRun model anymore. Today we experimented with removing it, and it seems to be a cleaner structure.

Instead of the JobRun endpoints, we tried making it a Dataset-related action: `POST /datasets/run_job/` with the following inputs:
- `original_dataset`: reference to an existing Dataset object, which will gain references to all new Dataset objects resulting from the job in its `derived_datasets` field, and which will be used as the source for the input image
- `region` (optional): A JSON specifying a polygon to use to extract a subimage from the whole of the original data (This is not yet implemented, currently stands as a TODO)
- `script`: reference to an existing JobScript object
- `additional_inputs` (optional): additional args for the job

This was inspired by the discussion yesterday about our JobRun solution possibly being redundant. Mary and I are hoping to finish experimenting with this restructure tomorrow morning. @waxlamp thoughts?
